### PR TITLE
Function-based storage control strategy

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address?
Addresses #102 (not completely resolved).

##### What are the significant changes in functionality due to this change request?
Adds the ability to define a `<strategy>` for `<stores>` Components and provide a Function to govern the dispatch decision making (regardless of the optimization). Note this function will likely require advanced HERON understanding of the variables in place to use effectively.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [x] 4. Automated Tests should pass.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [x] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: src/dispatch/pyomo_dispatch.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 from itertools import compress
 
 import numpy as np
+from numpy.lib.recfunctions import assign_fields_by_name
```
Comment (PaulTalbot-INL) [inline src/dispatch/pyomo_dispatch.py]: I think VSCode added this for me when I didn't need it. I'll remove it as part of addressing review comments.

**Comment by PaulTalbot-INL** [comment]
@dylanjm Binghui and I looked through this, and I think it's ready for review.

File: src/dispatch/pyomo_dispatch.py (hunk 1 above)
Comment (dylanjm) [reply src/dispatch/pyomo_dispatch.py]: Did you still want to remove this?

File: src/dispatch/pyomo_dispatch.py [hunk 2]
```diff
@@ -231,6 +249,15 @@ def dispatch_window(self, time, time_offset,
     m.Activity.initialize(m.Components, m.resource_index_map, m.Times, m)
     # constraints and variables
     for comp in components:
+      # components using a governing strategy (not opt) are Parameters, not Variables
+      # TODO should this come BEFORE or AFTER each dispatch opt solve?
+      # -> responsive or proactive?
+      intr = comp.get_interaction()
+      if intr.is_governed(): #intr.is_type('Storage') and intr.get_strategy() is not None:
```
Comment (dylanjm) [inline src/dispatch/pyomo_dispatch.py]: Looks like there is some left-over logic here. Please remove if not needed.

File: src/dispatch/pyomo_dispatch.py [hunk 3]
```diff
@@ -289,6 +316,47 @@ def dispatch_window(self, time, time_offset,
...
+    """
+    tol = 1e-4 # TODO user option
+    if old is None:
+      return False
+    converged = True
+    for comp in components:
+      intr = comp.get_interaction()
+      name = comp.name
+      if intr.is_governed():
+      #if intr.is_type('Storage') and intr.get_strategy() is not None:
```
Comment (dylanjm) [inline src/dispatch/pyomo_dispatch.py]: Might be worthwhile to remove this logic but keep the TODO comment.

**Review by dylanjm — CHANGES_REQUESTED** [review]
Everything looks good. A few minor changes.

**Review by dylanjm — APPROVED** [review]
Looks good. I'll go ahead and merge

**Pull request merged**